channels/jaicp/src/main/kotlin/com/justai/jaicf/channel/jaicp/JaicpConnector.kt

```diff
@@ -15,13 +21,59 @@ const val DEFAULT_PROXY_URL = "https://bot.jaicp.com"
  * @property botApi the [BotApi] implementation used to process the requests for all channels
  * @property channels is a list of channels which will be managed by connector
  * @property accessToken can be configured in JAICP Web Interface
- *
  * */
-interface JaicpConnector {
-    val botApi: BotApi
-    val channels: List<JaicpChannelFactory>
-    val accessToken: String
-    val url: String
+abstract class JaicpConnector(
+    val botApi: BotApi,
+    val channels: List<JaicpChannelFactory>,
+    val accessToken: String,
+    val url: String,
+    httpClient: HttpClient
+) : WithLogger {
+
+    private val chatAdapterConnector = ChatAdapterConnector(accessToken, url, httpClient)
+    private val registeredChannels = parseChannels()
+
+
+    protected fun registerChannels() {
+        registeredChannels.forEach { (factory, cfg) ->
+            when (factory) {
+                is JaicpCompatibleChannelFactory -> {
+                    registerChannel(factory.create(botApi), cfg)
+                        .also { logger.info("JAICP-compatible channel has been created for $it") }
+                }
+
+                is JaicpNativeChannelFactory -> {
+                    registerChannel(factory.create(botApi), cfg)
+                        .also { logger.info("JAICP-native channel has been created for $it") }
+                }
+
+                is JaicpCompatibleAsyncChannelFactory -> {
+                    registerChannel(factory.create(botApi, getChannelProxyUrl(cfg)), cfg)
+                        .also { logger.info("JAICP-compatible async channel has been created for $it") }
+                }
+                else -> logger.info("Channel type ${factory.channelType} is not added to list of channels in BotEngine")
+            }
+        }
+    }
+
+    private fun parseChannels(): List<Pair<JaicpChannelFactory, ChannelConfig>> {
+        val registeredChannels: List<ChannelConfig> = chatAdapterConnector.listChannels()
+        logger.info("Retrieved ${registeredChannels.size} channels configuration")
+
+        return channels.flatMap { factory ->
+            registeredChannels.mapNotNull {
+                if (factory.channelType.toUpperCase() == it.channelType) {
+                    factory to it
+                } else
+                    null
+            }
+        }
+    }
+
+    abstract fun registerChannel(channel: JaicpBotChannel, channelConfig: ChannelConfig)
+
+    protected fun getChannelProxyUrl(config: ChannelConfig) =
```

Comment:
Nobody overrides this

Reply:
It doesn't, but another subclass (JaicpPollingConnector) uses it to set proper url